### PR TITLE
Rework the interface of cranelift-module

### LINKED ISSUE #1691: Increase maximum alignment for data objects
#### Feature

Currently an `u8` is used to represent the alignment. This means that the biggest power of two that fits in the alignment specifier is 128 bytes. I propose to use at least an `u16` for alignment.

#### Benefit

A rustc test depends on 256 byte alignment. There is also code that requires much bigger alignments like 4096 bytes or even bigger.

#### Implementation

Replace `u8` with `u16` to specify alignment in `cranelift-module`.

#### Alternatives

None

### PR DESCRIPTION
**Pull request by bjorn3**
The old design has a `Module` struct that wraps around a `Backend` implementation. Because the `Backend` is not exposed it is not possible to export extra methods to do things like deallocating specific functions or adding new predefined imports for the jit. It is also not possible to customise the behavior of existing methods to for example allow replacing an already defined function for hot code swapping. It was also not possible to create a module with a non-static lifetime (required for headcrab-inject) In the new design `Module` is instead a trait with `ModuleDeclarations` being a struct for some shared behavior regarding how declarations are handled. As the `Module` is fully exposed, it is possible to add new methlds to the api of every module and customize the behavior of existing methods. None of the previously mentioned use cases are implemented yet. I will leave that for a future PR.

Also fixes #1691

### CODE REVIEW

**Comment by pchickey** [comment]
This looks cool - I want to spend some time reviewing it but I'm just about to leave for vacation, I'll get to it early next week.

File: cranelift/module/src/data_context.rs [hunk 1]
```diff
@@ -50,6 +50,8 @@ pub struct DataDescription {
     pub data_relocs: Vec<(CodeOffset, ir::GlobalValue, Addend)>,
     /// Object file section
     pub custom_segment_section: Option<(String, String)>,
+    /// Alignment
```
Comment (sunfishcode) [inline cranelift/module/src/data_context.rs:53]: Please document whether this is in bytes or in log2(bytes).

Comment (bjorn3) [reply cranelift/module/src/data_context.rs]: Done

**Comment by pchickey** [comment]
Sorry for the delay - this is excellent work. The only functionality missing that we need in Lucet is an iterator across all function declarations. I have added two iterator getters to ModuleDeclarations here: https://github.com/bytecodealliance/wasmtime/commit/13c6bdd9bab2e8aaacad420a2f90b87624b8e6f3

**Comment by bjorn3** [comment]
Added your commit to this PR.

**Review by pchickey — APPROVED** [review]

**Pull request merged**